### PR TITLE
Add Notification module for Local Notifications

### PR DESCRIPTION
**Pull request by satya164**
A simple API to manage local notifications. This is currently Android only. Eventually we can merge the PushNotificationIOS APIs into this (especially the local notification ones).

API:

```js
const n = Notification.show({
  title: 'Sticky notification with title, body and count',
  body: 'Notification body with some text',
  count: 3,
  link: 'http://facebook.github.io/react-native',
  sticky: true,
  vibrate: [200, 100, 200],
});

setTimeout(() => n.close(), 5000);

Notification.closeAll();
```

The API tries to be consistent with the the HTML5 notifications API in terms of naming of properties - https://developer.mozilla.org/en/docs/Web/API/notification.

Standard properties,
`body`, `sticky`, `slient`, `vibrate`, `tag`

Non-standard properties:
`count`, `priority`, `link`

Unsupported properties:
`dir`, `lang`, `icon`, `noscreen`, `renotify`, `sound`

cc @martinbigio @mkonicek @astreet

### CODE REVIEW

**Comment by satya164** [comment]
Something to consider is, if we want to be consistent with the HML5 notification API - https://developer.mozilla.org/en/docs/Web/API/notification

We won't be able to support lots of things it specifies, and will prolly want to support non-standard options. So not sure about it.

**Comment by satya164** [comment]
cc @vjeux @nicklockwood need inputs on the API

**Comment by nicklockwood** [comment]
FWIW, we already expose local notifications for iOS via the PushNotificationIOS module, but I'm not really happy with the iOS API because we don't currently expose an intuitive way to enumerate notifications, or cancel them individually.

I like the idea of sticking to the HTML5  spec, but this implementation seems to suffer from the same problem that we encountered on iOS, that returning a token to use for cancelling notifications only really works for the lifetime of the app. As soon as the app is restarted, the tokens are lost, but the notifications remain scheduled, so now you've lost any way to cancel them.

The way I envisioned solving this on iOS would be to enumerate the scheduled notifications when then app starts up, and then serialize them (or perhaps just their ids) as constantsToExport on the native side so that the JS module could expose them as an array of notification objects that could then be manipulated (cancelled, edited and rescheduled, etc).

I'm not sure if that fits with the way the HTML5 API works, but it seems more useful in practice than having orphaned notifications. But in any case, if we are going to create a new local notifications module we should probably do so with a view to making a common x-platform interface rather than something Android specific.

It's fine to land for Android first (in fact, I'd encourage it, since x-plat PRs are tricky to review), but lets make sure that the API can be implemented the same way on iOS as well.

Cc: @dmmiller

**Comment by satya164** [comment]
> The way I envisioned solving this on iOS would be to enumerate the scheduled notifications when then app starts up

This doesn't seem to be possible on Android. AFAIK you would need a service which can keep track of notifications.

**Comment by nicklockwood** [comment]
Looking at the HTML spec, it seems like it doesn't encompass the concept of scheduled notifications at all, right? You can simply show them and then hide them.

That is quite different from how Local Notifications work on iOS. We can show notifications immediately by setting the fire date to now, but we can't hide them again programatically as far as I'm aware.

So maybe it makes sense to have 3 APIs:

1) Notifications, that follows the HTML5 spec as closely as possible (minus hiding, on iOS)
2) PushNotifications, that handles server-sent async notifications
3) TimedNotifications, that handles scheduled notifications (and may be iOS specific, at least for now)

With the latter two APIs being built in the same basic style as the HTML5 spec (e.g same property naming conventions), but departing from it as appropriate to support the native requirements.

**Comment by satya164** [comment]
> Looking at the HTML spec, it seems like it doesn't encompass the concept of scheduled notifications at all, right? You can simply show them and then hide them.

Yup. 

> PushNotifications, that handles server-sent async notifications

This will just handle the events, right? When we receive a push notification, we can show it using the `Notification` API.

**Comment by vjeux** [comment]
Feel free to not feel restricted to use the web apis if it doesn't make sense. Especially for push notifications which no one uses on the web.

**Comment by bestander** [comment]
fyi buck build fails

**Comment by satya164** [comment]
@bestander Never worked with Buck. Tips to fix?

**Comment by bestander** [comment]
@satya164 it is easy, BUCK is very simple. :)
And it is good to start getting more into it because we are planning to make it a first class build system for RN because it is more predictable and super fast.

The idea of BUCK is acyclic trees.
A subtree in a file tree is managed by a BUCK file, every file in the folder structure that is at the same level with BUCK file or below can only be referred/compiled by this BUCK file.

In your case you are referring a JAVA file from managed by BUCK file in react/modules/something from another JAVA file managed by BUCK file in react/shell.
Just add a reference to react/modules/something in your react/shell BUCK file.
But beware of cycles, BUCK does not like cyclic dependencies.

We would appreciate if you fix this because we would need to do it anyway to import the code

**Comment by satya164** [comment]
@bestander Ah. Thanks for explaining. Definitely wanna know more about Buck :)

Will fix.

**Comment by janicduplessis** [comment]
@nicklockwood LocalNotification tokens could be saved to disk using localStorage to be able to cancel them across app launches.

I think we could stick to 2 APIs LocalNotification and PushNotification. LocalNotification would have support for delayed notifications that can be canceled using a token returned by the function that creates the notification. Seems doable on Android using AlarmManager.

**Comment by satya164** [comment]
> Seems doable on Android too using AlarmManager.

We would need a service to show the actual scheduled notification.

**Comment by mkonicek** [comment]
Thanks for adding the example! Do you have some screenshots? This shows Android system notifications, correct?

nit: `new Notification` showing a notification is unusual - constructors shouldn't have side effects (in any programming language I think). I would suggest adding a static method, something like `Notification.show` (don't know what the iOS API looks like, ideally it should be consistent where possible).

**Comment by satya164** [comment]
> Thanks for adding the example! Do you have some screenshots? This shows Android system notifications, correct?

Yeah. Will share screenshots in a bit.

> nit: new Notification showing a notification is unusual - constructors shouldn't have side effects (in any programming language I think). I would suggest adding a static method, something like Notification.show (don't know what the iOS API looks like, ideally it should be consistent where possible).

Yeah. That's just how the HML Notification API works. Happy to change it to Notification.show.

**Comment by satya164** [comment]
@mkonicek 

screenshot_20160213-214550

screenshot_20160213-214608

From lollipop onwards Android makes the icon monochrome (that's why the square). We'll need a way to specify the icon (Android only supports resource IDs as small icon).

**Comment by nicklockwood** [comment]
@satya164 

> This will just handle the events, right? When we receive a push notification, we can show it using the Notification API

Hmm, notifications are really very different on iOS. On iOS if you get a locla or push notification when the appis in the background it shows an alert or a "toast" (to use Android terminology) automatically.

But if you get a notification when the appis running, it happens silently - it's up to the app to display an alert or some other UI as appropriate. But (and here's where it really diverges from Android) there is no toast API on iOS - *only* the OS can show notification banners, and it only does so when it receives a notification.

So to build a notification API that works when the app is in the foreground on iOS, we'd have to build our own toast-like UI, and it most likely wouldn't look or behave like standard iOS notifications.

So basically, I'm not sure this API makes sense on iOS at all. In which case maybe just call it NotificationsAndroid for now and don't worry about x-platform at all. We can come up with an x-plat abstraction for the other kinds of notification later.

**Comment by nicklockwood** [comment]
@janicduplessis 

>LocalNotification tokens could be saved to disk using localStorage to be able to cancel them across app launches.

There's already a way to enumerate scheduled notifications using the UILocalNotifications API. I'd much rather use that than store a second representation, which might end up getting out of sync.

**Comment by satya164** [comment]
@nicklockwood Ah. Didn't know that. Yeah, it seems entirely different. I'll rename this to `NotificationAndroid`.

**Comment by janicduplessis** [comment]
@nicklockwood It was mostly so that it could work with Android too but I agree that what is possible to do is so different that it might not be worth it trying to make something cross platform here.

File: Libraries/Notification/NotificationAndroid.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ * To show a new statusbar notification, call `NotificationAndroid.show` with a `details` object,
+ *
+ * ```
+ * const notification = NotificationAndroid.show({
+ *   title: 'Test notification',
+ *   body: 'This notification will change your life'
+ * });
+ *
+ * // Close the notification after 5 seconds
+ * setTimeout(() => notification.close(), 5000);
```
Comment (dmmiller) [inline Libraries/Notification/NotificationAndroid.js]: should say notification.dismiss()

File: Libraries/Notification/NotificationAndroid.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/**
```
Comment (dmmiller) [inline Libraries/Notification/NotificationAndroid.js]: Should this be NotificationAndroid.android.js and should we add a .ios.js that is not implemented rather than just marking all the methods below as @platform android?

File: Libraries/Notification/NotificationAndroid.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+ * Copyright (c) 2015-present, Facebook, Inc.
```
Comment (dmmiller) [inline Libraries/Notification/NotificationAndroid.js]: Oh, actually this file should probably go away since you have a NotificationAndroid under Libraries/Notification and also one below under Libraries/NotificationAndroid

File: Libraries/NotificationAndroid/NotificationAndroid.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+   *
+   * @platform android
+   */
+  static show(details: NotificationDetails) {
+    invariant(
+      typeof details === 'object' && details !== null && typeof details.title === 'string',
+      'Details must be a valid object with a title property'
+    );
+
+    NotificationModule.presentLocalNotification(details, NOTIFICATION_ID);
```
Comment (dmmiller) [inline Libraries/NotificationAndroid/NotificationAndroid.js]: Should you use the same NOTIFICATION_ID every time?  Do you want to bump it each time?  If the tag is not specified you will end up having two notifications with the same (tag, id) pair which is not good.

File: Libraries/Notification/NotificationAndroid.js (hunk 3 above)
Comment (satya164) [inline Libraries/Notification/NotificationAndroid.js]: Oops. Sorry. Committed this by mistake. Will remove.

File: Libraries/NotificationAndroid/NotificationAndroid.js (hunk 4 above)
Comment (satya164) [inline Libraries/NotificationAndroid/NotificationAndroid.js]: @dmmiller 

> Should you use the same NOTIFICATION_ID every time? Do you want to bump it each time?
Using the same ID allows to replace/update a Notification. If the user wants to add a new one, he can add a tag.

> If the tag is not specified you will end up having two notifications with the same (tag, id) pair which is not good.
If the tag is same, the the previous notification will be replaced. So there won't be 2 notifications with tag and id pair.

If don't do this, there won't be a way to replace a previous notification. Or we can generate a random tag every time if we don't want the previous notification to be replaced by default (which feels ugly).

File: Libraries/Notification/NotificationAndroid.js (hunk 2 above)
Comment (satya164) [inline Libraries/Notification/NotificationAndroid.js]: @dmmiller Most of other components don't follow this approach (`ActionSheetIOS`), some do (`AppStateIOS`). But yeah, I think it'll be a good idea to have one as `.ios.js`.

**Comment by satya164** [comment]
@dmmiller Updated!

File: Libraries/NotificationAndroid/NotificationAndroid.js (hunk 4 above)
Comment (dmmiller) [inline Libraries/NotificationAndroid/NotificationAndroid.js]: The replace explanation seems fair.  If you want to have distinct, use a tag, otherwise calling it again will replace any existing one.  I'm okay with that.

**Comment by satya164** [comment]
Anything else I should change here?

**Comment by mkonicek** [comment]
@satya164 Was the purpose of this to use it for HMR or are you now using it in your apps for other things too?

I wonder how the module is going to pan out once someone adds support for push notifications. Do you think the API in this PR for local notifications is going to stay the same?

**Comment by satya164** [comment]
@mkonicek
 
> @satya164 Was the purpose of this to use it for HMR or are you now using it in your apps for other things too?

The original purpose was HMR, but we decided that we won't be using this for HMR. I'm not using it in my apps yet, mainly because currently my push notification code is Java only. But looking at @grabbou 's gist here - https://gist.github.com/grabbou/c25b836aacd58f724b9f, I should be able to use JS and this module.

> I wonder how the module is going to pan out once someone adds support for push notifications. Do you think the API in this PR for local notifications is going to stay the same?

Yeah, I think the API will stay the same. We will want to add some additional customizations like the style of the notification though.

Sorry for the late reply.

**Comment by mkonicek** [comment]
Thank you for the explanation @satya164! Should we hold off with this PR until it's used in production (one of main principles for code in fb repos)? What do you think?

**Comment by satya164** [comment]
@mkonicek Sure. I'll migrate my code to use this in next couple of weeks and will report back.

**Comment by facebook-github-bot-1** [comment]
@mkonicek would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by mkonicek** [comment]
@satya164 Do you plan to continue working on this? If not should we close the pull request to keep the number of pull requests under control and open a new one when you work on this?

**Comment by lacker** [comment]
This pull request seems like it's been put on hold for now, since no action in the past month. I'm going to close this pull request but please feel free to reopen this if you are intending to work on it again.
